### ISSUE TITLE
Proposed 3D Map enhancements

### ISSUE THREAD

**Issue by gregmalcolm** [EDDiscovery/EDDiscovery#39]
There are number of little enhancements I'd love to work on for the 3D map. I thought I'd list them to see if there's is any objection. The plan is to hit each small item as a separate pull request so that there's the opportunity to reject or request change individually. Also if we end up working on the same area at once it's easier to just throw away the proposed change and say "oh well!".

**3D Map:**
1) Autosave window size settings for map. So if I maximize it is always maximized. 
2) Default to zoomed out view showing whole map. I think the zoom level should be persisted as saved setting too. **Update: think the zoom should depend on star spasity. Existing default Zoom is fine in populated space, less useful in deep space.**
3) Button to max zoom out and recenter the rotation.
5) ~~Keyboard controls: I'd like cursor keys and ASWD for translating the map up down left and right. + and - for zoom.~~ **DONE!**
6) Mouse drag: Should behave like ED map itself. Dragging the mouse will cause the camera to lock on the mouse pointer and stay locked on it as the mouse moves, rotating on the Z Axis. 
7) Right button should climb and descend the Z axis
8) Bug: Center button sometimes stops working after some rotation. (Should also double check it doesn't break trilateration lines)
9) Grid: ~~Would be nice if there was a toggleable grid. Perhaps something in line with the in-game grid~~ **Looks like something along these lines is already in place. Although may still be fun to see if it can be improved on.**
10) Depth Perception: I want to experiment with lighting and star billboards (sprites) to give the sensation of some stars being further away than others. The sprites will need alpha information.
11) Toggleable Text Labels: Would be nice if key POIs like Sol, Sagittarius A, Beagle Point and maybe one or two others had text labels to make it clearer what we're looking at. The current 
position should also be labelled. Maybe zooming in causes more labeling detail.
12) Click on star to center on it and receive a little information about what it is.
13) Mouse hover to show individual star labels one at a time.
14) Make it so that when you access the 3D map it always goes to the same window. So you don't have multiple 3D maps floating around.
15) ~~3D Map should center on the system selected before clicking on "3D Map". This should also be the default "Center" textbox entry~~ - Thanks @iainross
16) Would love the ability to enter grid coordinates jumping to those locations
17) ~~Default button is becoming less relevant. Banish it to the settings page at some point?~~ - Thanks @iainross
18) When re-centering from the map it would be cool if a line was drawn from your current system and maybe distance information were added?

Some of the items near the end of the list are moving from "quick change" to "challenging, may not happen". But listing them because why not?

**Comment by finwen** [comment]
Great points.  But we might need to change render engine to get everything
we wants,

I can give you write access to the repository to if you wants.

2015-12-29 17:03 GMT+01:00 Greg Malcolm <notifications@github.com>:

**Comment by gregmalcolm** [comment]
Thanks, sounds good! Although if you don't mind I'd still be inclined to put most of my changes through as Pull Requests just so it gets some peer review.

I just took a glance at how you're rendering the map. OpenGL right? Should be pretty powerful... :D

**Comment by finwen** [comment]
I add you to the group.

So you can make smaller changes directly.

And a branch for bigger test  changes.

2015-12-29 17:20 GMT+01:00 Greg Malcolm <notifications@github.com>:

**Comment by finwen** [comment]
Yes openGL ashould be enough,   Just unsure if the .net wrapper i am using
have some limitations...

Its pretty fast to draw all points now atleast.

2015-12-29 17:27 GMT+01:00 Robert Wahlström <robert.l.wahlstrom@gmail.com>:

**Comment by gregmalcolm** [comment]
Ah. well we could always extend the wrapper if we have to if it's "open"... :D

**Comment by Red13x3** [comment]
I like all your suggestions.

The only real issue with the map is that it doesn't rotate and move around like the in-game one. It disorients me and makes my eyes "pop" to see the Galaxy "inverse" to what it should be, while I don't get that in game. (As in, the Thargoid bubble will be "towards me" and then when I rotate the map a little, it will suddenly "pop" and look like it is far from me. Hard to explain but it's a visual effect thing)

Mark
(CMDR Red_13)

--
Mark Jelic
iPhone: 0402 322 901

**Comment by gregmalcolm** [comment]
Yep, that's exactly the kind of thing area I want to improve on. I'd like rotation and translation to mimic the game control style to some degree.

**Comment by finwen** [comment]
I just started experimenting in the grid. Need to adapt it  too zoom level and decide a good standard color.

**Comment by gregmalcolm** [comment]
I actually tweaked that a little in my latest PR. Apologies if my change has messed up what you're working on. It got pretty involved... :D

**Comment by finwen** [comment]
Go ahead and work on it. I will concentrate my work in some other areas for
a while.  Like the new prospecting module.

2016-01-01 7:02 GMT+01:00 Greg Malcolm <notifications@github.com>:

**Comment by Red13x3** [comment]
Prospecting module would be very welcome! Only place to put that info now is in the Notes.

Btw, another thing I think would be interesting to track, for us explorers, is a checkbox indicating we were the first to discover that system or the first star or even just a planet. Check it on if you discovered it first. Leave it off, otherwise.

Mark

--
Mark Jelic
iPhone: 0402 322 901

**Comment by finwen** [comment]
Yes we should add a such button soon.

On the 3dMaps.  Your new grid color looks good.
Now we need ato adapt grid density with zooming. Should not be to hard to
do.

2016-01-01 11:06 GMT+01:00 Mark <notifications@github.com>:

**Comment by gregmalcolm** [comment]
Yep. I was thinking it would be useful to see the exact cursor positions on
screen to supplement eyeballing the grid. I started on it in the status bar
but as I wasn't going to be done before pushing up code I went ahead and
hid the status bar for now.

Another thought for the Future: encapsulate map in a User Control. Then
we'd still be able to open an external window, but could also split screen
the main page when the user clicks on a known system to show where it is
immediately. Not actually planning on doing it anytime soon, but it's
something I've been something that I've been pondering.

**Comment by gregmalcolm** [comment]
Just updated the list of things to do. I've changed quite a few to an updated opinion of what needs doing for quite a few of them.

**Comment by Red13x3** [comment]
As long as the 3D map works like the in-game on, as much as possible, I'll be happy. (In terms of using the mouse to rotate and translate and zoom in and out.)

**Comment by iainross** [comment]
re points 15 and 17 - I added the default centre because I want it it to open centred on the same place every time whatever I've got selected in the Travel History - just not Sol...

Noticed it opening on the selected system last night and thought there was a bug before I noticed it had been introduced deliberately.

So, rather than just "fixing" it back to the way I want it, I'm thinking I could add a setting that controls whether the map opens on the default centre system or the selected system (possibly extending 'selected system' so that if the grid selection doesn't have coordinates it looks for the nearest one - by number of jumps since we don't have coordinates so can't get an actual distance - in the the travel history that does have rather than it defaulting to Sol where the selected system has no coordinates) and a toggle on the map dialog to switch quickly between default centre and travel history selection.  The new setting could go on the settings tab and the setting of the default centre moved there at the same time.

Anyone object to that?

**Comment by Red13x3** [comment]
What you're suggesting sounds fine to me. 

Btw, I had a go at the latest build  that's available to download and the 3D map is coming along very well!

I did have a problem when I tried to move the centre of the rotation by using the two button mouse drag or even just the right mouse button drag. It basically messed the location up and had to close and re-open to fix it. 

--
Mark
iPhone: 0402 322 901
